File: backend/app/schemas/jobs.py
"""Wire models for the job-contract operator surface.

These answer the three questions the master plan's target state asks of the job
layer - **what ran, what failed, what it cost** - in the one canonical vocabulary.

Two deliberate choices worth stating, because both are the point rather than detail:

``succeeded`` is computed by ``app.jobs.status.is_success``, which returns True for
``completed`` and nothing else. The flag is on the wire so a UI cannot re-derive
success from a status string and get it wrong - there is exactly one definition, and
``degraded`` is not it.

``reason`` is never empty on a ``degraded`` or ``blocked`` run: the database's CHECK
constraint refuses to store one that is. So a client rendering a degraded run always
has something true to display, and "partially succeeded" with no explanation is not a
state this API can emit.

Python attributes are snake_case with a ``serialization_alias`` (ruff N815 forbids a
raw camelCase attribute), so the emitted JSON is camelCase like every other module
here and a future ``frontend/lib/jobs.ts`` lines up one-for-one.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.jobs.status import JobStatus, is_success, needs_attention


def _iso(value: Any) -> str | None:
    return value.isoformat() if isinstance(value, datetime) else None


def _duration_seconds(started: Any, finished: Any) -> float | None:
    if isinstance(started, datetime) and isinstance(finished, datetime):
        return round((finished - started).total_seconds(), 3)
    return None


class JobRunResponse(BaseModel):
    """One execution of one logical unit of background work."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    job_name: str = Field(serialization_alias="jobName")
    task: str
    queue: str
    status: JobStatus

    #: The ONE definition of success, computed server-side. See the module docstring.
    succeeded: bool
    #: True for degraded / blocked / failed - the three an operator must act on.
    needs_attention: bool = Field(serialization_alias="needsAttention")

    client_id: str | None = Field(default=None, serialization_alias="clientId")
    client_name: str = Field(default="", serialization_alias="clientName")
    scope_type: str = Field(default="", serialization_alias="scopeType")
    scope_id: str | None = Field(default=None, serialization_alias="scopeId")

    attempt: int
    max_attempts: int = Field(serialization_alias="maxAttempts")

    detail: str = ""
    #: Guaranteed non-empty when status is degraded or blocked (DB check constraint).
    reason: str = ""
    #: The machine-readable half of `reason`: a stable snake_case identifier an
    #: operator surface groups by. Also guaranteed non-empty for degraded/blocked.
    reason_code: str = Field(default="", serialization_alias="reasonCode")
    error_type: str = Field(default="", serialization_alias="errorType")
    error_message: str = Field(default="", serialization_alias="errorMessage")

    cost_usd: float = Field(default=0.0, serialization_alias="costUsd")

    correlation_id: str = Field(serialization_alias="correlationId")
    parent_run_id: str | None = Field(default=None, serialization_alias="parentRunId")
    idempotency_key: str | None = Field(default=None, serialization_alias="idempotencyKey")

    created_at: str | None = Field(default=None, serialization_alias="createdAt")
    started_at: str | None = Field(default=None, serialization_alias="startedAt")
    finished_at: str | None = Field(default=None, serialization_alias="finishedAt")
    heartbeat_at: str | None = Field(default=None, serialization_alias="heartbeatAt")
    scheduled_for: str | None = Field(default=None, serialization_alias="scheduledFor")
    cancel_requested: bool = Field(default=False, serialization_alias="cancelRequested")
    duration_seconds: float | None = Field(default=None, serialization_alias="durationSeconds")

    result: dict[str, Any] | None = None

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> JobRunResponse:
        status = str(row["status"])
        return cls(
            id=str(row["id"]),
            job_name=str(row["job_name"]),
            task=str(row.get("task") or ""),
            queue=str(row["queue"]),
            status=JobStatus(status),
            succeeded=is_success(status),
            needs_attention=needs_attention(status),
            client_id=str(row["client_id"]) if row.get("client_id") else None,
            client_name=str(row.get("client_name") or ""),
            scope_type=str(row.get("scope_type") or ""),
            scope_id=str(row["scope_id"]) if row.get("scope_id") else None,
            attempt=int(row.get("attempt") or 0),
            max_attempts=int(row.get("max_attempts") or 1),
            detail=str(row.get("detail") or ""),
            reason=str(row.get("reason") or ""),
            reason_code=str(row.get("reason_code") or ""),
            error_type=str(row.get("error_type") or ""),
            error_message=str(row.get("error_message") or ""),
            cost_usd=float(row.get("cost_usd") or 0),
            correlation_id=str(row["correlation_id"]),
            parent_run_id=str(row["parent_run_id"]) if row.get("parent_run_id") else None,
            idempotency_key=row.get("idempotency_key"),
            created_at=_iso(row.get("created_at")),
            started_at=_iso(row.get("started_at")),
            finished_at=_iso(row.get("finished_at")),
            heartbeat_at=_iso(row.get("heartbeat_at")),
            scheduled_for=_iso(row.get("scheduled_for")),
            cancel_requested=row.get("cancel_requested_at") is not None,
            duration_seconds=_duration_seconds(row.get("started_at"), row.get("finished_at")),
            result=row.get("result"),
        )


class DeadLetterResponse(BaseModel):
    """A unit of work the platform accepted and did not deliver."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    run_id: str | None = Field(default=None, serialization_alias="runId")
    job_name: str = Field(serialization_alias="jobName")
    task: str = ""
    queue: str
    client_id: str | None = Field(default=None, serialization_alias="clientId")
    client_name: str = Field(default="", serialization_alias="clientName")
    scope_type: str = Field(default="", serialization_alias="scopeType")
    scope_id: str | None = Field(default=None, serialization_alias="scopeId")
    correlation_id: str | None = Field(default=None, serialization_alias="correlationId")
    idempotency_key: str | None = Field(default=None, serialization_alias="idempotencyKey")

    attempts: int = 0
    #: Carried from the run so the queue can be grouped by CAUSE, not just by job name.
    reason_code: str = Field(default="", serialization_alias="reasonCode")
    error_type: str = Field(default="", serialization_alias="errorType")
    error_message: str = Field(default="", serialization_alias="errorMessage")
    #: The sanitized traceback. Staff-only; secrets are redacted before it is stored.
    traceback: str = ""
    #: Everything needed to replay: {"args": [...], "kwargs": {...}}.
    payload: dict[str, Any] = Field(default_factory=dict)

    dead_lettered_at: str | None = Field(default=None, serialization_alias="deadLetteredAt")
    first_failed_at: str | None = Field(default=None, serialization_alias="firstFailedAt")
    replayed_at: str | None = Field(default=None, serialization_alias="replayedAt")
    replayed_run_id: str | None = Field(default=None, serialization_alias="replayedRunId")
    resolved_at: str | None = Field(default=None, serialization_alias="resolvedAt")
    resolution: str = ""
    #: Still awaiting a human decision - the actual queue.
    open: bool = True

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> DeadLetterResponse:
        return cls(
            id=str(row["id"]),
            run_id=str(row["run_id"]) if row.get("run_id") else None,
            job_name=str(row["job_name"]),
            task=str(row.get("task") or ""),
            queue=str(row["queue"]),
            client_id=str(row["client_id"]) if row.get("client_id") else None,
            client_name=str(row.get("client_name") or ""),
            scope_type=str(row.get("scope_type") or ""),
            scope_id=str(row["scope_id"]) if row.get("scope_id") else None,
            correlation_id=str(row["correlation_id"]) if row.get("correlation_id") else None,
            idempotency_key=row.get("idempotency_key"),
            attempts=int(row.get("attempts") or 0),
            reason_code=str(row.get("reason_code") or ""),
            error_type=str(row.get("error_type") or ""),
            error_message=str(row.get("error_message") or ""),
            traceback=str(row.get("traceback") or ""),
            payload=dict(row.get("payload") or {}),
            dead_lettered_at=_iso(row.get("dead_lettered_at")),
            first_failed_at=_iso(row.get("first_failed_at")),
            replayed_at=_iso(row.get("replayed_at")),
            replayed_run_id=str(row["replayed_run_id"]) if row.get("replayed_run_id") else None,
            resolved_at=_iso(row.get("resolved_at")),
            resolution=str(row.get("resolution") or ""),
            open=row.get("resolved_at") is None and row.get("replayed_at") is None,
        )


class JobStatusCount(BaseModel):
    """One row of the summary: how many runs landed in a state, and what they cost."""

    model_config = ConfigDict(populate_by_name=True)

    status: JobStatus
    runs: int
    cost_usd: float = Field(serialization_alias="costUsd")
    succeeded: bool


class JobSummaryResponse(BaseModel):
    """The operator headline over a window.

    ``degraded`` is its own line, never folded into a success count - which is the
    entire reason the vocabulary distinguishes them.
    """

    model_config = ConfigDict(populate_by_name=True)

    window_hours: int = Field(serialization_alias="windowHours")
    by_status: list[JobStatusCount] = Field(serialization_alias="byStatus")
    total_runs: int = Field(serialization_alias="totalRuns")
    #: completed only.
    succeeded_runs: int = Field(serialization_alias="succeededRuns")
    #: degraded + blocked + failed.
    needs_attention_runs: int = Field(serialization_alias="needsAttentionRuns")
    total_cost_usd: float = Field(serialization_alias="totalCostUsd")
    open_dead_letters: int = Field(serialization_alias="openDeadLetters")

    @classmethod
    def from_rows(cls, rows: list[dict[str, Any]], *, window_hours: int, open_dlq: int) -> JobSummaryResponse:
        counts = [
            JobStatusCount(
                status=JobStatus(str(r["status"])),
                runs=int(r["runs"]),
                cost_usd=float(r.get("cost_usd") or 0),
                succeeded=is_success(str(r["status"])),
            )
            for r in rows
        ]
        return cls(
            window_hours=window_hours,
            by_status=counts,
            total_runs=sum(c.runs for c in counts),
            succeeded_runs=sum(c.runs for c in counts if c.succeeded),
            needs_attention_runs=sum(c.runs for c in counts if needs_attention(c.status)),
            total_cost_usd=round(sum(c.cost_usd for c in counts), 6),
            open_dead_letters=open_dlq,
        )


class InFlightResponse(BaseModel):
    """What the per-client concurrency cap is currently acting on."""

    model_config = ConfigDict(populate_by_name=True)

    client_id: str | None = Field(default=None, serialization_alias="clientId")
    client_name: str = Field(default="", serialization_alias="clientName")
    queue: str
    running: int

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> InFlightResponse:
        return cls(
            client_id=str(row["client_id"]) if row.get("client_id") else None,
            client_name=str(row.get("client_name") or ""),
            queue=str(row["queue"]),
            running=int(row["running"]),
        )


class CancelRequest(BaseModel):
    """Why an operator stopped a run. Recorded on the activity log, not the run row -
    the run row records that it was cancelled; the log records who and why."""

    model_config = ConfigDict(populate_by_name=True)

    reason: str = Field(default="", max_length=500)


class ResolveRequest(BaseModel):
    """Closing a dead letter requires saying what was decided.

    Enforced here AND by a CHECK constraint. A queue closed with no reasons written is
    a graveyard: the next person cannot tell "we fixed the underlying bug" from "we
    gave up on this one".
    """

    model_config = ConfigDict(populate_by_name=True)

    resolution: str = Field(min_length=1, max_length=1000)

    @field_validator("resolution")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("resolution cannot be blank")
        return value.strip()


class ReplayResponse(BaseModel):
    """The result of re-running a dead letter."""

    model_config = ConfigDict(populate_by_name=True)

    dead_letter_id: str = Field(serialization_alias="deadLetterId")
    run_id: str = Field(serialization_alias="runId")
    message_id: str = Field(serialization_alias="messageId")
    idempotency_key: str = Field(serialization_alias="idempotencyKey")

File: frontend/lib/jobs.ts
// ============================================================
// AIOS · Job contract — WIRE TYPES + display metadata.
// Mirrors backend `app/schemas/jobs.py` one-for-one: the Python attributes are
// snake_case with a `serialization_alias`, so the JSON is camelCase and drops
// straight into these types with no field mapping.
//
// THE VOCABULARY IS THE PRODUCT. Two rules this file exists to protect:
//
//   1. `succeeded` is computed SERVER-SIDE by `app.jobs.status.is_success`, which
//      is True for `completed` and NOTHING else. Never re-derive success from a
//      status string in a component — read the flag. `degraded` is not success:
//      a job that published two of ten pages did not keep its promise, and the
//      moment it is allowed to look like it did, the operator stops looking.
//   2. `reason` / `reasonCode` are GUARANTEED non-empty on `degraded` and
//      `blocked` (DB CHECK `job_runs_reason_required_ck`), so a degraded run
//      ALWAYS has a true explanation to render. Render it. "Partially succeeded"
//      with no reason is not a state this API can emit, and must not be a state
//      the board can display.
//
// Same for `needsAttention` (degraded | blocked | failed) — on the wire, computed
// once, not spelled out again per screen.
// ============================================================

// --- the ONE status vocabulary ----------------------------------------------
// Mirrors `public.job_status` / `app.jobs.status.JobStatus`. Seven words, every
// module reports in them, and the whole value is in the distinctions:
//   completed  the promise was kept
//   degraded   it finished but part of the promise was NOT kept
//   blocked    it deliberately did not spend (gate, missing credential, cap)
//   failed     it hit an error it could not recover from
//   cancelled  a human stopped it
export type JobStatus =
  | "queued"
  | "running"
  | "completed"
  | "degraded"
  | "blocked"
  | "failed"
  | "cancelled";

// --- one run of one logical unit of background work -------------------------
export type JobRun = {
  id: string;
  jobName: string;
  task: string;
  queue: string;
  status: JobStatus;

  /** The ONE definition of success, computed server-side. Do NOT re-derive it. */
  succeeded: boolean;
  /** True for degraded | blocked | failed — the three an operator must act on. */
  needsAttention: boolean;

  clientId: string | null;
  clientName: string;
  scopeType: string;
  scopeId: string | null;

  attempt: number;
  maxAttempts: number;

  detail: string;
  /** Non-empty whenever status is degraded or blocked (DB CHECK). Always show it. */
  reason: string;
  /** The machine-readable half of `reason`: a stable snake_case id to group by. */
  reasonCode: string;
  errorType: string;
  errorMessage: string;

  costUsd: number;

  /** Reassembles a fan-out: one nightly sweep and its 80 per-client jobs share it. */
  correlationId: string;
  parentRunId: string | null;
  idempotencyKey: string | null;

  createdAt: string | null;
  startedAt: string | null;
  finishedAt: string | null;
  /** The liveness signal the reaper reads — a stale heartbeat is a reclaimed run. */
  heartbeatAt: string | null;
  scheduledFor: string | null;
  /** Cancellation is COOPERATIVE: this flag is set, the job stops at a checkpoint. */
  cancelRequested: boolean;
  durationSeconds: number | null;

  result: Record<string, unknown> | null;
};

// --- a unit of work the platform ACCEPTED and did not deliver ---------------
// The most operationally important object on the board: it carries the payload,
// so the work is not lost — it is replayable.
export type DeadLetterPayload = {
  args?: unknown[];
  kwargs?: Record<string, unknown>;
};

export type DeadLetter = {
  id: string;
  runId: string | null;
  jobName: string;
  task: string;
  queue: string;
  clientId: string | null;
  clientName: string;
  scopeType: string;
  scopeId: string | null;
  correlationId: string | null;
  idempotencyKey: string | null;

  attempts: number;
  /** Carried from the run so the queue groups by CAUSE, not just by job name. */
  reasonCode: string;
  errorType: string;
  errorMessage: string;
  /** Sanitized traceback. Staff-only; secrets are redacted before it is stored. */
  traceback: string;
  /** Everything needed to replay it: {"args": [...], "kwargs": {...}}. */
  payload: DeadLetterPayload;

  deadLetteredAt: string | null;
  firstFailedAt: string | null;
  replayedAt: string | null;
  replayedRunId: string | null;
  resolvedAt: string | null;
  resolution: string;
  /** Still awaiting a human decision — the actual queue. */
  open: boolean;
};

// --- the operator headline over a window ------------------------------------
export type JobStatusCount = {
  status: JobStatus;
  runs: number;
  costUsd: number;
  /** Server-computed: true on the `completed` line and no other. */
  succeeded: boolean;
};

export type JobSummary = {
  windowHours: number;
  /** `degraded` is its own line here and is NEVER folded into a success count. */
  byStatus: JobStatusCount[];
  totalRuns: number;
  /** completed only. */
  succeededRuns: number;
  /** degraded + blocked + failed. */
  needsAttentionRuns: number;
  totalCostUsd: number;
  openDeadLetters: number;
};

// --- what the per-client concurrency cap is currently acting on -------------
// The answer to "why is that client's work not starting": a client at its cap on
// a queue has its next job deferred, then honestly `blocked` — never started.
export type InFlightRow = {
  clientId: string | null;
  clientName: string;
  queue: string;
  running: number;
};

// --- the result of re-running a dead letter ---------------------------------
export type ReplayResult = {
  deadLetterId: string;
  runId: string;
  messageId: string;
  /** `replay:<deadLetterId>` — deliberately NOT the original key, or the replay
   *  would find the old terminal run, skip the work, and look like it succeeded. */
  idempotencyKey: string;
};

// Aliases under the backend model names, so an import written against either
// vocabulary compiles. Same types, no second definition to drift.
export type JobRunResponse = JobRun;
export type DeadLetterResponse = DeadLetter;
export type JobSummaryResponse = JobSummary;
export type InFlightResponse = InFlightRow;
export type ReplayResponse = ReplayResult;

// --- filters for GET /jobs/runs ---------------------------------------------
// Query aliases exactly as the router declares them (status/jobName/clientId/
// correlationId/needsAttention + the shared limit/offset page window, capped at
// 200 server-side).
export type JobRunFilters = {
  status?: JobStatus;
  jobName?: string;
  clientId?: string;
  correlationId?: string;
  /** The view that matters day to day: every terminal run that was not clean. */
  needsAttention?: boolean;
  limit?: number;
  offset?: number;
};

// ---------------------------------------------------------------------------
// Display metadata
// ---------------------------------------------------------------------------
// FOUR tones, and the mapping is load-bearing:
//   ok    the promise was kept        → completed, and nothing else
//   warn  a human must decide         → degraded, blocked
//   bad   it broke                    → failed
//   idle  no verdict (yet, or ever)   → queued, running, cancelled
// `degraded` and `blocked` deliberately DO NOT read as ok. That is the entire
// reason the vocabulary distinguishes them from `completed`.
export type JobTone = "ok" | "warn" | "bad" | "idle";

export type JobStatusMeta = {
  label: string;
  tone: JobTone;
  /** Material Symbols name, matching the rest of the dashboard's meta maps. */
  icon: string;
};

export const JOB_STATUS_META: Record<JobStatus, JobStatusMeta> = {
  queued:    { label: "Queued",    tone: "idle", icon: "schedule" },
  running:   { label: "Running",   tone: "idle", icon: "autorenew" },
  completed: { label: "Completed", tone: "ok",   icon: "check_circle" },
  // Finished, promise partly unkept. Amber, never green — and never without its reason.
  degraded:  { label: "Degraded",  tone: "warn", icon: "warning" },
  // Deliberately did not spend (gate / missing credential / cap). Not an error, not a win.
  blocked:   { label: "Blocked",   tone: "warn", icon: "block" },
  failed:    { label: "Failed",    tone: "bad",  icon: "error" },
  cancelled: { label: "Cancelled", tone: "idle", icon: "cancel" },
};

/** Display order for the summary strip: what worked, then what needs a human,
 *  then what is still moving. */
export const JOB_STATUSES: JobStatus[] = [
  "completed",
  "degraded",
  "blocked",
  "failed",
  "running",
  "queued",
  "cancelled",
];

const TERMINAL_STATUSES: ReadonlySet<string> = new Set<string>([
  "completed",
  "degraded",
  "blocked",
  "failed",
  "cancelled",
]);

/** True when nothing further will happen to the run (mirrors `jobs.status.TERMINAL`).
 *  Used to stop polling a finished run — NOT a success test; use `run.succeeded`. */
export function isTerminalStatus(status: JobStatus | string): boolean {
  return TERMINAL_STATUSES.has(String(status));
}

/**
 * Label + tone + icon for a status. TOLERANT by construction: the wire types it
 * as an enum, but `api.get<T>` is an unchecked cast, so a status this build has
 * never heard of must render as an honest neutral chip rather than crash the
 * board (the lesson `providerMeta` in lib/cost.ts was written to remember).
 */
export function statusMeta(status: JobStatus | string): JobStatusMeta {
  const known = (JOB_STATUS_META as Record<string, JobStatusMeta | undefined>)[String(status)];
  if (known) return known;
  const raw = String(status || "").trim();
  return {
    label: raw ? raw[0].toUpperCase() + raw.slice(1).replace(/[_-]+/g, " ") : "Unknown",
    // Neutral, not ok: an unrecognized state has told us nothing, least of all success.
    tone: "idle",
    icon: "help",
  };
}

/**
 * A run's wall-clock duration, from `durationSeconds` (computed server-side from
 * started_at/finished_at). Returns "" for a run that has not finished, so the
 * caller renders its own placeholder rather than a fake "0s".
 */
export function formatDuration(seconds: number | null | undefined): string {
  if (seconds === null || seconds === undefined || !Number.isFinite(seconds)) return "";
  const s = Math.max(0, seconds);
  if (s < 1) return `${Math.round(s * 1000)}ms`;
  if (s < 10) return `${s.toFixed(1)}s`;
  if (s < 60) return `${Math.round(s)}s`;
  if (s < 3600) {
    const m = Math.floor(s / 60);
    const rest = Math.round(s % 60);
    return rest ? `${m}m ${rest}s` : `${m}m`;
  }
  const h = Math.floor(s / 3600);
  const m = Math.round((s % 3600) / 60);
  return m ? `${h}h ${m}m` : `${h}h`;
}

/**
 * "5m ago" / "in 2h" from an ISO instant. BOTH directions on purpose: a run's
 * `scheduledFor` is in the future, and rendering it as "ago" would be a lie about
 * work that has not happened yet. "" for null/empty/unparseable, so a missing
 * timestamp renders as nothing rather than "Invalid Date".
 */
export { relativeTime } from "@/lib/format";

// --- who may act -------------------------------------------------------------
// Reads need `view_reports` (all six staff roles hold it; a portal client does
// NOT and is 403'd out of the namespace). Cancel / replay / resolve are LEAD
// actions — owner | admin | manager — because replay in particular deliberately
// re-runs work that already spent money once.
//
// The server is the boundary; this is the UX half, so a button that cannot work
// does not invite the click. Compared lowercase: /me serialises a Title-Case role
// while the backend permission check is lowercase.
export const JOB_LEAD_ROLES = ["owner", "admin", "manager"] as const;

export function isLeadRole(role: string | null | undefined): boolean {
  return JOB_LEAD_ROLES.includes(
    String(role ?? "").toLowerCase() as (typeof JOB_LEAD_ROLES)[number],
  );
}
